### PR TITLE
Functional Tests

### PR DESCRIPTION
**Pull request by brannow**
create new multicolumn container functional test

### CODE REVIEW

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php [hunk 1]
```diff
@@ -25,10 +25,14 @@
 
 use TYPO3\CMS\Core\Core\Bootstrap;
 use TYPO3\CMS\Core\DataHandling\DataHandler;
-use TYPO3\CMS\Core\Tests\FunctionalTestCase;
+use TYPO3\CMS\Core\Tests\Functional\DataHandling\AbstractDataHandlerActionTestCase;
 
-class tx_multicolumn_tcemainTest extends FunctionalTestCase
+class tx_multicolumn_tcemainTest extends AbstractDataHandlerActionTestCase
```
Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: Please do not use the DataHandler TestCase. It is an internal abstraction which is not needed for the tests in this extension. We don't want to test the DataHandler but the hook which are called from it.

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php [hunk 2]
```diff
@@ -47,33 +51,32 @@ protected function setUp()
         $fixturePath = ORIGINAL_ROOT . 'typo3conf/ext/multicolumn/Tests/Functional/Fixtures/';
         $this->importDataSet($fixturePath . 'tt_content.xml');
 
-        $this->setUpBackendUserFromFixture(1);
-        Bootstrap::getInstance()->initializeLanguageObject();
+        // we don't need this stuff as a child of AbstractDataHandlerActionTestCase anymore, the parent will do
+        //$this->setUpBackendUserFromFixture(1);
+        //Bootstrap::getInstance()->initializeLanguageObject();
     }
```
Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: See above. And if it's not needed anymore, please remove it ;-)

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php [hunk 3]
```diff
@@ -47,33 +51,32 @@ protected function setUp()
...
 
     /**
+     * copy an existing multicolumn container into another ROOT-PageColumn (see: tt_content.xml)
```
Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: just:
Copy an existing multicolumn container to another page

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php [hunk 4]
```diff
@@ -82,19 +85,19 @@ public function copyContainerAndChildrenToOtherColumnInDefaultLanguage()
 
         $count = $this->getDatabaseConnection()->exec_SELECTcountRows(
             '*',
-            'tt_content',
+            self::TABLE_CONTENT,
             'uid=3'
             . ' AND pid=1'
             . ' AND deleted=0'
-            . ' AND CType=\'multicolumn\''
+            . ' AND CType=\''.self::CTYPE_MULTICOLUMN.'\''
```
Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: psr-1 -> spaces before and after dot

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php [hunk 5]
```diff
@@ -106,28 +109,26 @@ public function copyContainerAndChildrenToOtherColumnInDefaultLanguage()
     }
 
     /**
+     * copy an existing multicolumn container into another page (see: tt_content.xml)
```
Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: copy an existing multicolumn container to another page

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php [hunk 6]
```diff
@@ -106,28 +109,26 @@ public function copyContainerAndChildrenToOtherColumnInDefaultLanguage()
...
-                ],
+            self::TABLE_CONTENT => [
+                1 => [
+                    'copy' => [
+                        'action' => 'paste',
+                        'target' => 2,
+                        'update' => [
+                            'colPos' => 0,
+                            'sys_language_uid' => 0
+                        ]
```
Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: Please insert semicolons after each item

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php [hunk 7]
```diff
@@ -136,19 +137,19 @@ public function copyContainerAndChildrenToOtherPageInDefaultLanguage()
 
         $count = $this->getDatabaseConnection()->exec_SELECTcountRows(
             '*',
-            'tt_content',
+            self::TABLE_CONTENT,
             'uid=3'
             . ' AND pid=2'
             . ' AND deleted=0'
-            . ' AND CType=\'multicolumn\''
+            . ' AND CType=\''.self::CTYPE_MULTICOLUMN.'\''
```
Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: psr-1

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php [hunk 8]
```diff
@@ -159,4 +160,37 @@ public function copyContainerAndChildrenToOtherPageInDefaultLanguage()
         $this->assertSame(1, $count);
     }
 
+    /**
+     * create an new Multicolumn container
+     *
+     * @test
+     */
+    public function createNewContainerWithDefaultLanguage()
```
Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: CreateNewContainerInDefaultLanguage

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php [hunk 9]
```diff
@@ -159,4 +160,37 @@ public function copyContainerAndChildrenToOtherPageInDefaultLanguage()
         $this->assertSame(1, $count);
     }
 
+    /**
+     * create an new Multicolumn container
```
Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: Create an new multicolumn container

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php [hunk 10]
```diff
@@ -159,4 +160,37 @@ public function copyContainerAndChildrenToOtherPageInDefaultLanguage()
...
+     * @test
+     */
+    public function createNewContainerWithDefaultLanguage()
+    {
+        $config = [
+            'header' => 'Mutlicolumn insert Test',
+            'colPos' => 0,
+            'CType' => self::CTYPE_MULTICOLUMN,
+            'sys_language_uid' => 0,
+            'tx_multicolumn_parentid' => 0
```
Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: Please remove this -> should be either set by the hook or fallback to default field configuration

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php [hunk 11]
```diff
@@ -159,4 +160,37 @@ public function copyContainerAndChildrenToOtherPageInDefaultLanguage()
...
+
+        if (isset($result[self::TABLE_CONTENT])) {
+            foreach ($result[self::TABLE_CONTENT] as $ttcId) {
+                $count = $this->getDatabaseConnection()->exec_SELECTcountRows(
+                    '*',
+                    self::TABLE_CONTENT,
+                    'uid='.$ttcId
+                    . ' AND pid=1'
+                    . ' AND deleted=0'
+                    . ' AND CType=\''.self::CTYPE_MULTICOLUMN.'\''
```
Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: psr-1

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php (hunk 6 above)
Comment (brannow) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: you mean 'comma', or?

Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: right

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php [hunk 12]
```diff
@@ -52,111 +55,167 @@ protected function setUp()
     }
 
     /**
+     * get the current UID from the LAST copied parent UID
+     *
+     * @param $oldUid
+     * @return int
+     */
+    protected function getUIDFromCopy($oldUid)
```
Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: This function is not needed. You can use $dataHandler->copyMappingArray[$table][$origUid]

File: Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php [hunk 13]
```diff
@@ -52,111 +55,167 @@ protected function setUp()
...
             . ' AND CType=\'textpic\''
             . ' AND colPos=10'
-            . ' AND tx_multicolumn_parentid=3'
+            . ' AND tx_multicolumn_parentid=' . $containerUid
         );
         $this->assertSame(1, $count);
     }
 
+    /**
+     * Create an new multicolumn container
```
Comment (IchHabRecht) [inline Tests/Functional/Hooks/class.tx_multicolumn_tcemainTest.php]: Create a new multicolumn container

File: Tests/Functional/Fixtures/tt_content.xml [hunk 14]
```diff
@@ -18,4 +18,13 @@
 		<sys_language_uid>0</sys_language_uid>
 		<tx_multicolumn_parentid>1</tx_multicolumn_parentid>
 	</tt_content>
+	<tt_content>
```
Comment (IchHabRecht) [inline Tests/Functional/Fixtures/tt_content.xml]: This element is not needed in this patch/test, is it?

Comment (brannow) [inline Tests/Functional/Fixtures/tt_content.xml]: not in this test, yes

**Comment by IchHabRecht** [comment]
Closed with #10

**Pull request closed without merging**